tracing/opencensus/grpc.go

```diff
@@ -0,0 +1,156 @@
+package opencensus
+
+import (
+	"context"
+
+	"go.opencensus.io/trace"
+	"go.opencensus.io/trace/propagation"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/metadata"
+	"google.golang.org/grpc/status"
+
+	kitgrpc "github.com/go-kit/kit/transport/grpc"
+)
+
+const propagationKey = "grpc-trace-bin"
+
+// GRPCClientTrace enables OpenCensus tracing of a Go kit gRPC transport client.
+func GRPCClientTrace(options ...TracerOption) kitgrpc.ClientOption {
+	cfg := TracerOptions{
+		sampler: trace.AlwaysSample(),
+	}
+
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	clientBefore := kitgrpc.ClientBefore(
+		func(ctx context.Context, md *metadata.MD) context.Context {
+			var name string
+
+			if cfg.name != "" {
+				name = cfg.name
+			} else {
+				name = ctx.Value(kitgrpc.ContextKeyRequestMethod).(string)
+			}
+
+			span := trace.NewSpan(
+				name,
+				trace.FromContext(ctx),
+				trace.StartOptions{
+					Sampler:  cfg.sampler,
+					SpanKind: trace.SpanKindClient,
+				},
+			)
+
+			if !cfg.public {
+				traceContextBinary := string(propagation.Binary(span.SpanContext()))
+				(*md)[propagationKey] = append((*md)[propagationKey], traceContextBinary)
+			}
+
+			return trace.NewContext(ctx, span)
+		},
+	)
+
+	clientFinalizer := kitgrpc.ClientFinalizer(
+		func(ctx context.Context, err error) {
+			if span := trace.FromContext(ctx); span != nil {
+				s, ok := status.FromError(err)
+				if ok {
+					span.SetStatus(trace.Status{Code: int32(s.Code()), Message: s.Message()})
+				} else {
+					span.SetStatus(trace.Status{Code: int32(codes.Unknown), Message: err.Error()})
+				}
+				span.End()
+			}
+		},
+	)
+
+	return func(c *kitgrpc.Client) {
+		clientBefore(c)
+		clientFinalizer(c)
+	}
+
+}
+
+// GRPCServerTrace enables OpenCensus tracing of a Go kit gRPC transport server.
+func GRPCServerTrace(options ...TracerOption) kitgrpc.ServerOption {
+	cfg := TracerOptions{}
+
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	serverBefore := kitgrpc.ServerBefore(
+		func(ctx context.Context, md metadata.MD) context.Context {
+			var (
+				spanContext trace.SpanContext
+				ok          bool
```

Comment:
Nit: given the distance between declaration and use, can this have a slightly more descriptive name?